From 21393e64b9524b89d373d2d303afce38919c96c2 Mon Sep 17 00:00:00 2001
From: Julien Danjou <julien.danjou@enovance.com>
Date: Thu, 6 Oct 2011 17:15:32 +0200
Subject: [PATCH 27/44] Include original exception in ClassNotFound exception

By doing this, we allow the error messages to be more useful. When an import
of a class fails because of a missing module dependency, it would have fail
that way for example:

$ nova-manage
Traceback (most recent call last):
  File "./bin/nova-manage", line 84, in <module>
    from nova import image
  File "/home/jd/Work/src/nova/nova/image/__init__.py", line 22, in <module>
    from nova.image import glance
  File "/home/jd/Work/src/nova/nova/image/glance.py", line 42, in <module>
    GlanceClient = utils.import_class('glance.client.Client')
  File "/home/jd/Work/src/nova/nova/utils.py", line 66, in import_class
    raise exception.ClassNotFound(class_name=class_str)
nova.exception.ClassNotFound: Class Client could not be found

This does not help the user, since it indicates the class Client cannot be
found, even if it is actually found but fail to import.

With this commit, the error message is better:
nova-manage
Traceback (most recent call last):
  File "./bin/nova-manage", line 84, in <module>
    from nova import image
  File "/home/jd/Work/src/nova/nova/image/__init__.py", line 22, in <module>
    from nova.image import glance
  File "/home/jd/Work/src/nova/nova/image/glance.py", line 42, in <module>
    GlanceClient = utils.import_class('glance.client.Client')
  File "/home/jd/Work/src/nova/nova/utils.py", line 66, in import_class
    raise exception.ClassNotFound(class_name=class_str, exception=exc)
nova.exception.ClassNotFound: Class Client could not be found: No module named kombu.connection

This helps to know that in this kombu is missing.

It would probably even better to rename ClassNotFound to
ClassCannotBeImported or something like that too.

(cherry picked from commit 33e58925c75b66c6800d3ba6068ee1e4f0db6617)

Change-Id: I50c591b54820ae734e5d1dccedf6653c5c9d40f9
Signed-off-by: Julien Danjou <julien.danjou@enovance.com>
---
 Authors           |    1 +
 nova/exception.py |    2 +-
 nova/utils.py     |    2 +-
 3 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/Authors b/Authors
index adb2bbc..cc78535 100644
--- a/Authors
+++ b/Authors
@@ -63,6 +63,7 @@ Josh Durgin <joshd@hq.newdream.net>
 Josh Kearney <josh@jk0.org>
 Josh Kleinpeter <josh@kleinpeter.org>
 Joshua McKenty <jmckenty@gmail.com>
+Julien Danjou <julien.danjou@enovance.com>
 Justin Santa Barbara <justin@fathomdb.com>
 Justin Shepherd <jshepher@rackspace.com>
 Kei Masumoto <masumotok@nttdata.co.jp>
diff --git a/nova/exception.py b/nova/exception.py
index 79b1bf3..b88e8e5 100644
--- a/nova/exception.py
+++ b/nova/exception.py
@@ -727,7 +727,7 @@ class NetworkAdapterNotFound(NotFound):
 
 
 class ClassNotFound(NotFound):
-    message = _("Class %(class_name)s could not be found")
+    message = _("Class %(class_name)s could not be found: %(exception)s")
 
 
 class NotAllowed(NovaException):
diff --git a/nova/utils.py b/nova/utils.py
index 81157a4..c64b11f 100644
--- a/nova/utils.py
+++ b/nova/utils.py
@@ -63,7 +63,7 @@ def import_class(import_str):
         return getattr(sys.modules[mod_str], class_str)
     except (ImportError, ValueError, AttributeError), exc:
         LOG.debug(_('Inner Exception: %s'), exc)
-        raise exception.ClassNotFound(class_name=class_str)
+        raise exception.ClassNotFound(class_name=class_str, exception=exc)
 
 
 def import_object(import_str):
-- 
1.7.6.5

